src/metaswitch/clearwater/config_manager/test/test_config_access.py

```diff
@@ -1031,6 +1045,41 @@ def test_missing_args(self):
         with self.assertRaises(SystemExit):
             config_access.parse_arguments()
 
+    def test_wrong_config(self, mock_loadplugins):
+        """check that when a config_type not in the choice list is provided
+        a error is raised"""
+
+        sys.argv = ["config_access.py",
+                    "download",
+                    "weatherwax_config",
+                    "--management_ip", "1.2.3.4",
+                    "--site", "siteX",
+                    "--etcd_key", "lockpick"]
+
+        with self.assertRaises(SystemExit):
+            config_access.parse_arguments()
+
+
+@mock.patch("metaswitch.clearwater.config_manager.config_access.load_plugins_in_dir",
+            return_value=[dns_json_config_plugin.DnsJson('path'),
+                          bgcf_json_config_plugin.BgcfJson('path'),
+                          shared_config_config_plugin.SharedConfig('path')])
+class TestClassReturn(unittest.TestCase):
```

Comment:
You've written two tests in this class, but I don't think the second one is adding very much. Fundamentally it will run through the exact same underlying code logic each time.

This brings up an interesting question: how do we decide which tests are useful? The way I like to approach this is by asking: "What different types of inputs could this function receive?". Then once you've identified all the different types, you can write a test for each. To see how this works, let's go through this function as an example:

`class_return()` takes two parameters - `key` and `filepath`. 
- For the `key`, there are a bunch of possible inputs, but they will all be treated in one of two ways: either they will find the matching `ConfigType` class or they won't. Therefore, we should test both of these cases. 
- For the `filepath`, we expect that this will be passed into the `ConfigType` class somehow. We should test that it is passed through correctly.
- We should also think about whether there's any interaction between the inputs. In this case, I don't think there is.

The function also calls through to `load_plugins_from_dir()`. We should have think about what sort of inputs we could get from there, because `load_plugins_from_dir()` behaves differently depending on the state of the system. `class_return()` will behave the same way for most things it's possible for `load_plugins_from_dir()` to return. But there is one potentially interesting case - what happens when it returns nothing? Will this cause issues? We should probably check.

So from this, we get a list of four test cases we should run:
- valid key
- invalid key
- filepath set correctly on ConfigType object
- `load_plugins_from_dir()` returns no classes

Note that we can arrive at this list with limited understanding of what the actual code inside the function looks like. Furthermore, if you were to write these four tests, you would end up finding a bug in your code - it doesn't handle invalid keys correctly!

By taking this approach to unit tests, you can end up proactively finding bugs in your code. If you don't do it like this, you can sometimes find that writing the UTs ends up devolving into a box-ticking exercise.